### PR TITLE
Make target-spec json file extensions case-insensitive

### LINKED ISSUE #127387: Support target json file extension in various cases (json, JSON)
Currently user's target json file can have only `json` form, but for file that ends with `.JSON' rustc returns error:
```
error: Error loading target specification: Could not find specification for target "my-target.JSON". Run `rustc --print target-list` for a list of built-in targets
```

Where it useful?

Old OS and old file-systems such as FAT (that stored the files names and extensions as all caps).

### Meta

`rustc --version --verbose`:
```
rustc 1.81.0-nightly (ba1d7f4a0 2024-06-29)
binary: rustc
commit-hash: ba1d7f4a083e6402679105115ded645512a7aea8
commit-date: 2024-06-29
host: aarch64-apple-darwin
release: 1.81.0-nightly
LLVM version: 18.1.7
```

### PR DESCRIPTION
**Pull request by boozook**
Currently user's target json file can have only `json` ext, but cannot have `JSON' or any other strange form.

This is a little fix with added support of case-insensitive file-ext for user specified target only.
That means that not supported in following cases:
- any directory traversal, such as in `$RUST_TARGET_PATH`
- same for `rustc_target::spec::Target::search`, so it doesn't work with `JSON` on case-sensitive FS, but __works fine on case-insensitive FS__.
- `$RUST_TARGET_PATH` dir traversal in the bootstrap is not touched too.

## Performance cost:

Almost without changes.
Comparison using `eq_ignore_ascii_case` for __only latest ext_len chars__ of the `OsStr` (as bytes).

## Testing

How to test:
1. `rustc --print target-spec-json -Zunstable-options > ./some-target.JSON`
1. remove `is-builtin` field from the `some-target.JSON` or set to `false`
1. `rustc --print target-spec-json -Zunstable-options --target=some-target.JSON` should work
1. `rustc --print target-spec-json -Zunstable-options --target=some-target` should work on case-insensitive FS
1. `rustc --print cfg --target=some-target` should work too, on case-insensitive FS

Tests added.

## Implementation note

I've separated the comparison into a method for potential future use instead of just `str.ends_with(".json")` and to maintain compatibility. 
The method is public for the same reasons.

Fixes #127387 issue.

try-job: test-various
try-job: armhf-gnu
try-job: x86_64-msvc
try-job: aarch64-apple

### CODE REVIEW

File: compiler/rustc_target/src/spec/mod.rs [hunk 1]
```diff
@@ -3709,6 +3709,18 @@ impl TargetTriple {
...
+    /// Does not check if the path exists.
+    pub fn has_json_ext<S: AsRef<str>>(triple: S) -> bool {
+        const EXT: &'static str = ".json";
+        triple.as_ref().ends_with(EXT) || {
+            let s = triple.as_ref();
+            s.get(s.len().wrapping_sub(EXT.len())..)
+                .map(|ext| ext.eq_ignore_ascii_case(EXT))
+                .unwrap_or_default()
+        }
+    }
```
Comment (tgross35) [inline compiler/rustc_target/src/spec/mod.rs:3723]: You could just let `Path` do the work here, `Path::new(triple).extension().is_some_and(|ext| ext.eq_ignore_ascii_case(EXT))`.

But maybe better just to name this `has_ext_ignore_case` and take the desired extension as an argument? So it could be reused with other files, (probably doesn't need to be in `TargetTriple` either since there's nothing triple-specific).

**Comment by tgross35** [comment]
Thanks for the fix, and welcome!

> because I suppose no one but me will cry if it ever breaks, and this is minor change and doesn't changes existing behaviour, so existing tests with .json are passed. Should I add tests?

If you're up for it it's never a bad thing to have a regression test, even if it means you'll be the only one no longer crying :) You can just translate your test instructions into a [rmake test](https://github.com/rust-lang/rust/tree/524d806c62a82ecc0cf8634b94997ae506f4d6f9/tests/run-make) (mirror a test that uses the newer `rmake.rs`, the `Makefile` tests are going away).

File: compiler/rustc_target/src/spec/mod.rs (hunk 1 above)
Comment (boozook) [reply compiler/rustc_target/src/spec/mod.rs]: Thank you for reviewing.

> You could just let Path do the work here, Path::new(triple).extension().is_some_and(|ext| ext.eq_ignore_ascii_case(EXT)).

I tried to optimise, so in case of ext is `"json"` it costs exactly same as before my patch. 
But now I see, double cost if there is no `".json"` 🤔 You totally right, so for your suggestion in approximately middle case cost will be a little bit higher then before the patch, but definitely not double.

Comment (boozook) [reply compiler/rustc_target/src/spec/mod.rs]: Done.


So we have this: (actual for any input)
 - `file_name`: `components().next_back().and_then(/* closure with match */)`
 - `extension`: `file_name` + `map(rsplit_file_at_dot).and_then(|(opta, optb)| opta.and(optb))`

Imho, in some future could be good to optimise it like following:
 - take slice of bytes by the input
 - take slice of last {len of ext} chars
   - if can't => false
 - `eq_ignore_ascii_case` for that slice and ext
 It should be slightly better in cost.

However, the price is so low that with your permission, I might look into it later with a new PR.

File: tests/run-make/target-specs-filename/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// This test checks the rustc can accept target-spec-json with cases
+// including file extension - it could be in upper and lower case.
+// Used to test: print target-spec-json and cfg
+// Issue: https://github.com/rust-lang/rust/issues/127387
+
+///@ ignore-cross-compile
```
Comment (boozook) [inline tests/run-make/target-specs-filename/rmake.rs:6]: @tgross35 ☝🏻 I'm not sure should it be here or I can remove ignore-cross-compile?

File: tests/run-make/target-specs-filename/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+fn main() {
+    // This is a matrix [files x args], files for target, args for print.
+    for file in [
+        "normal-target-triple",
+        "normal-target-triple.json",
+        "ext-in-caps.JSON",
+        "ALL-IN-CAPS.JSON",
+    ] {
+        for args in [["cfg"].as_slice(), &["target-spec-json", "-Zunstable-options"]] {
+            let output = rustc().arg("--print").args(args).args(["--target", file]).run();
```
Comment (jieyouxu) [inline tests/run-make/target-specs-filename/rmake.rs:17]: Suggestion: sorry I know there's currently no docs for the support library (still working on it), but `.run()` asserts that the output status is succecssful so you don't need the status assert ^^

File: tests/run-make/target-specs-filename/rmake.rs (hunk 2 above)
Comment (jieyouxu) [reply tests/run-make/target-specs-filename/rmake.rs]: Since you don't try to run a built binary, this should be able to handle cross-compilation. But we don't have to assume, we can run try-jobs to check!

File: tests/run-make/target-specs-filename/rmake.rs (hunk 3 above)
Comment (boozook) [reply tests/run-make/target-specs-filename/rmake.rs:17]: I've realized it couple minutes ago. Thanks, I'll remove it.

File: tests/run-make/target-specs-filename/rmake.rs (hunk 2 above)
Comment (jieyouxu) [reply tests/run-make/target-specs-filename/rmake.rs]: For context, I'm running your test on 4 CI runners: `test-various` and `armhf-gnu` CI runners in particular has cross-compilation setups that will exercise your test in cross-compilation context.

File: tests/run-make/target-specs-filename/rmake.rs (hunk 3 above)
Comment (boozook) [reply tests/run-make/target-specs-filename/rmake.rs:17]: Done ❤️‍🔥

**Comment by boozook** [comment]
Should I push just one commit instead of these few?

File: tests/run-make/target-specs-filename/rmake.rs (hunk 2 above)
Comment (boozook) [reply tests/run-make/target-specs-filename/rmake.rs]: Thank you, that's cool!

**Comment by jieyouxu** [comment]
> Should I push just one commit instead of these few?

When being reviewed, it's perfectly fine to have intermediate commits, but if you think it helps review or otherwise makes your commits more logical/cleaner, feel free to group logical changes into individual commits and rebase/force-push if you need to. A reviewer may ask you to reorganize the commit history before merge to tidy it up.

**Comment by boozook** [comment]
Oops, I just merged master into this by mistake.

@jieyouxu, thanks for explanation. I'll rebase it after bors complete tests. I [need to know](https://github.com/rust-lang/rust/pull/127389#discussion_r1667377103) is my test ok for cross-compiled compiler.

**Comment by boozook** [comment]
Latest force-push by me is just rebase + squash all into one commit.

It's hope bors is ok after this.

File: tests/run-make/target-specs-filename/rmake.rs (hunk 2 above)
Comment (boozook) [reply tests/run-make/target-specs-filename/rmake.rs]: As you said, [all works](https://github.com/rust-lang/rust/pull/127389#issuecomment-2211836260) without `ignore-cross-compile`. Bors competed test job 🎉

**Comment by jieyouxu** [comment]
Okay cool, the impl and the test case LGTM, but since this technically changes user-facing observable behavior of a stable flag, I'll leave it to Wesley to approve.

**Review by jieyouxu — APPROVED** [review]

**Comment by epage** [comment]
Unfamiliar with T-compiler care abouts but the original issue has no discussion and I don't see any MCP to this

Generally, cargo has held off on supporting alternative cases for files on case-sensitive file systems.  We have limited case-insensitivity checking for improving error messages (e.g. "error: no Cargo.toml found but cargo.toml exists").  **Is a policy on this something we should coordinate on?**

**Comment by jieyouxu** [comment]
I'm not sure if this change needs an MCP (it might to be safe) or if this change can be made at all, which is why I left the decision to Wesley. But coordinating how case sensitivity is handled between rustc and cargo is definitely a good idea.

**Comment by ChrisDenton** [comment]
I don't think `Cargo` can even work on systems that uppercase all file names. Maybe rustc can but I wouldn't be surprised if it breaks something?

**Comment by jieyouxu** [comment]
Hmm, actually, now that I think about it more, I don't even think we have any CI runners that have a case insensitive filesystem, so I think the test wouldn't be exercised in the environment that the change is supposed to affect?

**Comment by ChrisDenton** [comment]
Yes, it's not the case-insensitive bit that I was asking about it's this from the issue:

> Where it useful?
>
> Old OS and old file-systems such as FAT (that stored the files names and extensions as all caps).

I've not looked at the PR but for really old FAT it did only have uppercase 8.3 names (up to eight filename characters and 3 extension characters) which is definitely not workable for Cargo and I don't think rustc alone would manage it either.

**Comment by jieyouxu** [comment]
Thanks for the context. @boozook I apologize if I came across as "this PR looks ready to merge and we can accept this change", but given that the use case seems very niche. I'm not sure if we *can* accept this change at all given Cargo can't handle it, and I'm not sure we want to support old OS or filesystems or even if it would work for rustc. I'll let Wesley who's way more knowledgeable determine this :3

**Comment by boozook** [comment]
@ehuss 

> Both Windows and macOS are case-insensitive.

By default, but default != standard. MacOs has buildin support for case sensitive fs and it works perfect.

@ChrisDenton There is many many cases actually, starting from real DOS FAT, ending with [that](https://superuser.com/questions/1297658/folder-names-become-uppercase-when-syncing-to-fat32-drive#1297675) like cases.

**Comment by boozook** [comment]
@jieyouxu 

> all given Cargo can't handle it

I'm working on it: https://github.com/rust-lang/cargo/pull/14197

**Comment by boozook** [comment]
By the way, I really don't understand what the cargo has to do with it? Can't I use the compiler without a cargo? Even my cargo patches will not be accepted, this change could be useful for the compiler anyway, and will improve its support on other systems, despite the fact that the price of this support is minimal (a couple of lines of changes and fewer repetitions are the same `.ents_with(.json)`).

Where am I really wrong?

**Comment by jieyouxu** [comment]
> Where am I really wrong?

It's not you who's wrong, it's my bad for giving conflicting signals. I just feel like I'm lacking knowledge about the implications of this change, which is why I'm handing the review back over to Wesley who is the compiler team co-lead.

**Comment by tgross35** [comment]
@boozook mind detailing the use case a bit since it seems like the exact details to support aren't clear? Are you just passing around target specs on old flash drives, or actually building on them :)

> By the way, I really don't understand what the cargo has to do with it? Can't I use the compiler without a cargo? Even my cargo patches will not be accepted, this change could be useful for the compiler anyway, and will improve its support on other systems, despite the fact that the price of this support is minimal (a couple of lines of changes and fewer repetitions are the same `.ents_with(.json)`).

Yes you absolutely can, everyone is just talking about consistency - it would be a less smooth user experience if some things break on uppercase extensions but others work fine. That's not to say that making cargo accept case insensitive files _has_ to happen in order for `rustc` to accept them, just that it's worth a bit of effort to make sure the entire use case is actually covered.

**Comment by boozook** [comment]
@tgross35,
> Yes you absolutely can, everyone is just talking about consistency - it would be a less smooth user experience if some things break on uppercase extensions but others work fine. That's not to say that making cargo accept case insensitive files has to happen in order for rustc to accept them, just that it's worth a bit of effort to make sure the entire use case is actually covered.

I absolutely agree with you. That's why I've made similar patch for cargo. And step by step this all will be better.
I started from rustc because it's kinda independent part of toolchain. Cargo is depends on rustc, so cargo is second step for me. 

@jieyouxu, Plz don't worry, that's really not your problem. That's me created PR without any issue approval by the team. I assumed that a small patch that doesn't break anything wouldn't be burdensome for anyone.

**Comment by boozook** [comment]
@epage 
> Generally, cargo has held off on supporting alternative cases for files on case-sensitive file systems. We have limited case-insensitivity checking for improving error messages (e.g. "error: no Cargo.toml found but cargo.toml exists"). Is a policy on this something we should coordinate on?

Imho, that is heavily sad, that I can't use any name of package manifest if I explicitly passing it like `--manifest-path=new-carGo.TOML`. But that's my opinion and of course I can be wrong.

**Comment by boozook** [comment]
So, what's next steps? Should I close this PR or wait some?

**Comment by tgross35** [comment]
If you see support for case insensitivity going beyond a one-off fix here, it is probably good to just file a Major Change Proposal (MCP) to propose a specific policy change. This is [an issue template at the compiler team repo](https://github.com/rust-lang/compiler-team/issues/new/choose). That will give you a chance to explain what use case we are currently not supporting well, which specific changes will help improve it, and collect input about it.

Please be patient; nobody is saying that this change is not desired or that you should close this PR. Its scope just needs to be better understood if it's not _only_ target spec files that are causing problems. (Which the MCP should help clarify).

**Comment by boozook** [comment]
@jieyouxu,
> I don't even think we have any CI runners that have a case insensitive filesystem, so I think the test wouldn't be exercised in the environment that the change is supposed to affect?

Let me explain it. This patch touches only that part where validation of user- or cargo- input via cli-arg or env, __but not any fs exercises like dir-traversal__. So before this patch I can't do `rustc --target=FILE.JSON`, but with the path I can. Nothing more. For example if I have in the pwd file `FILE.JSON` __on the case-sensitive fs__ and try `rustc --target=FILE` I'll get error because rustc will try to resole it as `FILE.json`. But that will be ok without error on the case-INsensitive fs.

**Comment by jieyouxu** [comment]
> So, what's next steps? Should I close this PR or wait some?

I would recommend waiting a bit to allow Wesley to provide you with better advice on which further steps to take.

**Comment by davidtwco** [comment]
r? @davidtwco

Apologies that this got lost in the cracks.

I think it's reasonable if niche that we would want to support accepting arbitrary filenames here (with whatever case or extension is provided, as long as we can find the file and parse it). It makes sense for our behaviour here to be consistent, so I would like to make sure that this is feasible anywhere else that we accept filenames from users, and I want to check if @rust-lang/cargo have any concerns with that other than the consistency between cargo/rustc?

If they don't, then we can file a MCP to agree on our policy going forward and then see this merged.

**Comment by epage** [comment]
>  I want to check if @rust-lang/cargo have any concerns with that other than the consistency between cargo/rustc?

My primary concern was us weighing whether consistency is important.

Would this change to rustc be prescriptive for cargo?  Whether Cargo should be consistent within itself is another question that would need to be considered and if there is an expectation for Cargo to match then we'd need to block on that decision.

Out of curiosity, I just tried to `mod HELLO;` a `HELLO.RS` and that was not supported.  So it seems like moving forward with this would be inconsistent within rustc.  I recognize that is on you all to decide.

**Comment by weihanglo** [comment]
> My primary concern was us weighing whether consistency is important.

Same here. Not only cargo and rustc but toolchain-wise including rustup, clippy and rustfmt. So I would say a policy where we accept where not

**Comment by ChrisDenton** [comment]
Thinking about this, I do wonder why rustc needs to care about the file extension at all.

Should we not just have a `--target-spec` option, then the file name (including the extension) can be anything the user wants it to be.

**Comment by epage** [comment]
The main reason I see for an extension check is to allow the format to be inferred if a new one is added in the future.

Us validating `--manifest-path` allowed cargo-script to be added and the early cargo team didn't predict that feature.

Another way to frame this is that the current behavior is consistent and we can always change it in the future but it a one-way door that can also constrain future decisions.

**Comment by epage** [comment]
Also, its easy to support case insensitive files when passed in but if a tool wants to support detecting the file, it requires scanning the whole directory.

**Comment by ChrisDenton** [comment]
Hm, I think this is a case where I would draw a distinction between rustc and Cargo. I think rustc almost never cares about any part of the file name, even the `.rs` is merely a strong convention. Whereas Cargo is "opinionated" (in the good sense), e.g. `Cargo.toml` has to be spelled exactly like that.

If we want to carve out space for future formats we could have `--target:json=path/to/json` or something like that.

**Comment by davidtwco** [comment]
> My primary concern was us weighing whether consistency is important.
> 
> Would this change to rustc be prescriptive for cargo? Whether Cargo should be consistent within itself is another question that would need to be considered and if there is an expectation for Cargo to match then we'd need to block on that decision.

That's a good point, they wouldn't necessarily need to be consistent. I think I'd be okay if they weren't. If we're okay with that then it's just a matter of whether t-compiler have an opinion here.

**Comment by boozook** [comment]
[Cargo PR](https://github.com/rust-lang/cargo/pull/14197) as second part of this story, waits this PR.

- - -

This branch looks stale. Should I update or rebase it?

**Comment by jieyouxu** [comment]
> This branch looks stale. Should I update or rebase it?

Probably wait until T-compiler discusses this change in today's triage meeting, the conflict doesn't really block the discussions.

**Comment by apiraino** [comment]
Discussed during T-compiler triage [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-12-12/near/488658633).

* To the concern of inconsistencies between rustc and other tooling, it seems we already have some of them ([comment](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-11-28/near/484946758)), though having rustc and cargo diverge on this is something we need to consider
* we are not sure if the question about the [actual usecase](https://github.com/rust-lang/rust/pull/127389#issuecomment-2214598978) was fully replied, more context would be helpful
* Targets like Cortex A were mentioned in [comment](https://github.com/rust-lang/rust/issues/127387#issuecomment-2213645373). Can we have some more details about which exactly ([this is our list](https://doc.rust-lang.org/rustc/platform-support.html) of supported compile targets)

In short, in general we would be in favor of this change, and we would also be happy with an [MCP](https://forge.rust-lang.org/compiler/mcp.html) documenting and detailing a bit more the intended scope of this change.

(please feel free to add more relevant questions from from the meeting)

@rustbot label -i-compiler-nominated

**Comment by epage** [comment]
> > My primary concern was us weighing whether consistency is important.
> > Would this change to rustc be prescriptive for cargo? Whether Cargo should be consistent within itself is another question that would need to be considered and if there is an expectation for Cargo to match then we'd need to block on that decision.
> 
> That's a good point, they wouldn't necessarily need to be consistent. I think I'd be okay if they weren't. If we're okay with that then it's just a matter of whether t-compiler have an opinion here.

@davidtwco I wanted to check in as my point could have been missed and yours is ambiguous enough.   I'm wondering if merging this PR is prescriptive to the Cargo team for merging rust-lang/cargo#14197.

- Should rustc and cargo agree on case-(in)sensitivity for target spec files?
- If yes, there is the question of self-consistency within Cargo
- If no, how much of a use case will be left for rustc to support this if Cargo chooses not to support this?

**Comment by davidtwco** [comment]
> > > My primary concern was us weighing whether consistency is important.
> > > Would this change to rustc be prescriptive for cargo? Whether Cargo should be consistent within itself is another question that would need to be considered and if there is an expectation for Cargo to match then we'd need to block on that decision.
> > 
> > 
> > That's a good point, they wouldn't necessarily need to be consistent. I think I'd be okay if they weren't. If we're okay with that then it's just a matter of whether t-compiler have an opinion here.
> 
> @davidtwco I wanted to check in as my point could have been missed and yours is ambiguous enough. I'm wondering if merging this PR is prescriptive to the Cargo team for merging [rust-lang/cargo#14197](https://github.com/rust-lang/cargo/pull/14197).
> 
> * Should rustc and cargo agree on case-(in)sensitivity for target spec files?
> * If yes, there is the question of self-consistency within Cargo
> * If no, how much of a use case will be left for rustc to support this if Cargo chooses not to support this?

I think it's okay for rustc to accept more than Cargo does. I'm entirely happen for Cargo to be more opinionated than rustc in circumstances like this.

> Discussed during T-compiler triage [on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-12-12/near/488658633).
> 
> * To the concern of inconsistencies between rustc and other tooling, it seems we already have some of them ([comment](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-11-28/near/484946758)), though having rustc and cargo diverge on this is something we need to consider
> * we are not sure if the question about the [actual usecase](https://github.com/rust-lang/rust/pull/127389#issuecomment-2214598978) was fully replied, more context would be helpful
> * Targets like Cortex A were mentioned in [comment](https://github.com/rust-lang/rust/issues/127387#issuecomment-2213645373). Can we have some more details about which exactly ([this is our list](https://doc.rust-lang.org/rustc/platform-support.html) of supported compile targets)

I'm not sure this is exactly what we agreed - we're happy to accept a change that checks for target files without the mandatory `.json` extension, it seems unnecessary that we mandate a specific extension, we should accept any filename users give us. If we wanted to go further than that with respect to case insensitivity, then we'd need to know more about the use cases that this addresses.

**Comment by ehuss** [comment]
> it seems unnecessary that we mandate a specific extension

I'm not sure I understand this. If we start supporting TOML spec files, would the compiler use a heuristic to determine which format it is? It seems unlikely that we would just drop JSON. I'm a little confused here.

**Comment by wesleywiser** [comment]
It's a good question but I think that's a slightly different issue. The rustc book currently [says](https://doc.rust-lang.org/rustc/targets/custom.html#custom-target-lookup-path):

> When rustc is given an option --target=TARGET (where TARGET is any string), it uses the following logic:
>
> 1. if TARGET is the name of a built-in target, use that
> 2. if TARGET is a path to a file, read that file as a json target
> 3. otherwise, search the colon-seperated list of directories found in the RUST_TARGET_PATH environment variable from left to right for a file named TARGET.json.
>
> These steps are tried in order, so if there are multple potentially valid interpretations for a target, whichever is found first will take priority. If none of these methods find a target, an error is thrown.

The discussion in the compiler team triage meeting concluded that if you write `--target foo.JSON`, that should just be accepted. Arguably, it should already be accepted because of point 2 in the described process. What we don't want is to support modifying point 3 to accept `TARGET.JSON` or any variation thereof.

If we decide to add TOML support (and it seems like that should have been the correct format to use anyway), point 2 will have to be modified to look at the file extension or something. Obviously, that again raises the question of the extension case but our consensus was that if you told rustc to look for `foo.JSON`, that should be ok since the user explicitly told us what the filename is. In the same way, a check for "is this JSON or TOML" should not be case-sensitive if the user already told us the proper filename. However, if the user leaves it up to us to infer the correct filename (by not specifying the full file path, eg point 3) then we will not go probing for variations in filename (or extension) casing.

**Comment by davidtwco** [comment]
> > it seems unnecessary that we mandate a specific extension
> 
> I'm not sure I understand this. If we start supporting TOML spec files, would the compiler use a heuristic to determine which format it is? It seems unlikely that we would just drop JSON. I'm a little confused here.

We could try parse it as JSON and then fall back to TOML if that fails.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
